cache/disk/findmissing.go

```diff
@@ -47,27 +62,43 @@ func (c *diskCache) FindMissingCasBlobs(ctx context.Context, blobs []*pb.Digest)
 		}
 
 		numMissing := c.findMissingLocalCAS(chunk)
-		if numMissing > 0 && c.proxy != nil {
+		if numMissing == 0 {
+			continue
+		}
+
+		if c.proxy != nil {
 			wg.Add(numMissing)
 			for i := range chunk {
-				if chunk[i] != nil {
-					c.containsQueue <- proxyCheck{
-						wg:     &wg,
-						digest: &chunk[i],
-						ctx:    ctx,
+				if chunk[i] == nil {
+					continue
+				}
+
+				if chunk[i].SizeBytes > c.maxProxyBlobSize {
+					// The blob would exceed the limit, so skip it.
+					wg.Add(-1) // remove from the waitgroup
+					if contextCancel != nil {
+						contextCancel()
 					}
+					continue
+				}
+
+				c.containsQueue <- proxyCheck{
+					wg:          &wg,
+					digest:      &chunk[i],
+					ctx:         ctx,
+					onProxyMiss: contextCancel,
 				}
 			}
+		} else if failFast {
+			break
 		}
 	}
 
 	if c.proxy != nil {
 		wg.Wait()
 	}
```

Comment:
hmmm.. If we are failing fast with a proxy, then there really is no reason to wait around. The context will be cancelled, so any pending `proxyCheck`'s will do no real work.

I've got a few thoughts while re-reading this code.
1. We should be checking `ctx.Done()` right before attempting to add to `conteainsQueue` since additions could block when the channel is full.
2. We should differentiate between the requestor cancelling `ctx` (returns a grpc Cancelled error) and our canceling in an effort to fast fail (not an error).
3. We should be watching the context(s) while we are waiting for the waitgroup to clear out. This will allow us to respect cancellation via the upstream context, fail fast, *and* let things spin down normally.